### DISCUSSION TITLE
How to select Count Property as y-axis for Histogram Plots?

### DISCUSSION THREAD

**Discussion by redundancer** [oxyplot/oxyplot#1872] [Q&A] [answered]
I have the following code:

```
private HistogramSeries CreateNewHistogramSeries(string title, double[] data, int numberOfBins, BinningOptions binningOptions)
        {
            var series = new HistogramSeries { Title = title };
            var bins = HistogramHelpers.CreateUniformBins(Enumerable.Min(data),
                Enumerable.Max(data), numberOfBins);
            var items = HistogramHelpers.Collect(data, bins, binningOptions);
            series.Items.AddRange(items);
            return series;
        }
```

When I put multiple of these series into a `new PlotModel()`, it suddenly uses the `Height` property instead of the `Count` property as y-axis. Why is this and how can I change it back to use the `Count` property as y-axis?

Here some HistogramItems:

![grafik](https://user-images.githubusercontent.com/81110847/161272309-ee074e7f-5c42-41e4-8585-cf6344a378f9.png)

And here how the plot looks like:

![grafik](https://user-images.githubusercontent.com/81110847/161272985-38b7ce85-dda6-4227-8087-04115be896e8.png)

It takes the `Height` property as y-axis which is not the behaviour I expect.

**Comment by VisualMelon** [comment, marked as answer]
`HistogramSeries` always uses `Height`, computed from `Area`. It ignores the `Count` for plotting. If you re-project the bins so that the `Area` is `Count * Width` that should give you what you want, but note that this only makes sense if all histogram items across all histograms use the same bin width.

**Reply by redundancer** [reply]
Thank you! This seems very counter intuitive to me. Is it not the usual expectation to have the number of occurrences in each bin as y-axis?

**Reply by VisualMelon** [reply]
For a bar chart, it would be, but for a histogram it is the area that is the meaningful value (bins concern a continuous range, rather than a discrete category).

**Comment by philipaxer** [comment]
I also ran across this. This really not very intuitive and leads to situations where frequency can be >1. 
In my opinion y-axis should show the relative frequency of how many items are in a specific bin (thus should always be within 0 and 1) or show the absolute count.

**Reply by VisualMelon** [reply]
Again, there are issues with interpretability with non-uniform bin widths, but I'd certainly be open to our supporting this scenario better as clearly there is some demand for it. Suggestions for API additions welcome. My first instinct would be to make it configurable whether to plot the dimensionally correct value or one where the total width under the graph is implicitly normalised to 1.

To achieve the sum being 1, and per the other answer, you'd need to re-project the bins so that the `Area` is `Count * Width / TotalCount` to achieve this at present (which is the same as `Area * Width`). Certainly the naming isn't ideal here: I'm regretting adding the `Value` property at all. (And sorry for all the edits: kept getting confused by what you were after: again, it doesn't make sense in general)

**Reply by philipaxer** [reply]
With "dimensionally correct value" you mean that bin_width*Y = count, correct? I dont think there is one "correct" way, and the other proposals are dimensionally incorrect. 
It depends what you want on the Y-axis. Typically this is absolute count or relative count (wrt the total number of samples). Normalizing Y to the bin size makes sense if the bin size is non-constant and you still want to compare the y-values in a meansing full way.

So I guess weare currently talking about 3 semantics: 
* Absolute
* Relative normalized to # samples, ignoring bin width
* Relative to the # of samples  and the bin width

I come from a Python background, and cannot really give good/solid suggestions for C# right now.  I personally think it is the job of the HistorgramSeries to present a HistrogramItem as the user wants - but that is probably preference.
API-wise, i find it weird to give HistogramItem a lot of knowledge (like Area). I think HisogramItem should only have count and the bin width (start/end). It is the Serie's job to present the thing as desired. 

But having said this, i dont know to much about the API philosophy here.

regards
Philip

**Reply by VisualMelon** [reply]
Sorry, didn't wish to imply I expected a proposal from you: just wanted to invite input from anyone.

To explain the thinking a little, for a simple scenario `HistogramItem` is overkill, but it gives flexibility where it is needed: there's no obligation to use the helper methods which are intended to support the simple case. This is not to say that it couldn't be configurable in the `HistogramSeries` how it is plotted. We could have an enum property to switch between plotting the `Count`, `Area`, and `Height` (which I think correspond to what you meant with your 3 bullet points): would that suit your scenario?

Doesn't help that the documentation is impossible to find still (my fault)

**Reply by philipaxer** [reply]
I still think that for complicated scenarios, HistogramItem is overkill. It only needs  a single property `value` and this is directly what is plotted by the Series.
It is then the task of the Helper or the user to populate this  `value` correctly. The Helper can get an enum which selects one of the three options i previously pointed out or you put there what ever you like. At the time when Item is created, all information that is ever needed is available.

It feels akward to replicated information with the intention that perhaps someone might need it later.

But again, that is probably preference where you want to have complexity. 

Matplotlib, for example, only takes a "scale" value per bin. That is the value that ends up on the plot. How you come up with scale is your job.

**Reply by VisualMelon** [reply]
I'd be disinclined to make the binning helpers configurable in this regard (unless we add even more meaningful properties to `HistogramSeries` and lean on the `Value` or `Height` for presentation independent of all other properties, but that feels dodgy). The numbers are readily fudgeable (demands awareness of the user), but with the existing design it doesn't make sense to make this part of the supported flow because it would too easily lead to flow. We could introduce new helpers that enforce uniform bins: that would avoid the main issues with this.

If this was supported then at present I'd suggest it should be on `HistogramSeries` (aligns well with the OP, which is asking how to configure the choice of Y axis value: I do see this as a presentation concern). The documentation therefore can make it clear that it only makes sense for uniform bin widths within and between series. Main concern would be what to do with the default tracker format string (probably leave it as is). Also need to check that this sort of configuration makes sense generally: it would have to be treated as a change of data.

**Reply by VisualMelon** [reply]
@Jonarw I think `HistogramSeries` predates your involvement here (I could be wrong, but was one of my earliest contributions) so interested to hear your opinion.

**Reply by VisualMelon** [reply]
Example plots showing the three options:
![image](https://github.com/oxyplot/oxyplot/assets/5646475/5f10617f-370a-4efc-a810-db0ff8f00faf)

Left is current behaviour (marginal proportion of whole, `height = area / width`)
Middle is the count (`count`)
Right is the proportion (`area`)

Code at https://gist.github.com/VisualMelon/ec1c6cb255344e26a88eb569cf622bf6 (uses example from existing example library)

Question is whether and how to support the middle and right possibilities in a non-hideous way.

**Reply by Jonarw** [reply]
Personally, I am most familiar with the Middle variant (showing the count), which I also find the most intuitive. This is not to say that the other variants don't have their uses, maybe they are more widely used in other fields. So I guess ideally all scenarios would be supported equally.

I remember that producing the Count variant is slightly awkward with the current API - you have to calculate the area by count * binwidth, only for the `HistogramItem` to divide by binwidth again during rendering. Not a big deal, and works well enough, but I understand that people could be confused by it.

I kind of like the idea of `HistogramItem` being 'dumber' - it really only needs start, end and value to be rendered. The rest could be sorted out by the user, who we could support with some helper methods.

But I also don't see this too critical: The current API does allow for all scenarios, some are just a bit more awkward than others.
